csrc/scheduler/matmul.cpp

```diff
@@ -46,14 +53,35 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
 
 } // namespace
 
-void scheduleMatmul(
-    TensorView* c,
-    TensorView* a,
-    TensorView* b,
-    MatmulParam& params) {
-  // Unpack from params.
-  auto& mma_builder = params.mma_builder;
-  auto& gemm_tile = params.tile_sizes;
+void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
+  const auto& inputs = fusion->inputs();
+  const auto& outputs = fusion->outputs();
+
+  TORCH_INTERNAL_ASSERT(
+      inputs.size() == 2,
+      "scheduleMatmul supports only fusions with two inputs");
+  TORCH_INTERNAL_ASSERT(
+      outputs.size() == 1,
+      "scheduleMatmul supports only fusions with single output");
+
+  TORCH_INTERNAL_ASSERT(
+      inputs[0]->isA<TensorView>(),
+      "fusion's first inpus is not an instance of TensorView class");
+  TORCH_INTERNAL_ASSERT(
+      inputs[1]->isA<TensorView>(),
+      "fusion's second inpus is not an instance of TensorView class");
+  TORCH_INTERNAL_ASSERT(
+      outputs[0]->isA<TensorView>(),
+      "fusion's output is not an instance of TensorView class");
+
+  TensorView* a = inputs[0]->as<TensorView>();
+  TensorView* b = inputs[1]->as<TensorView>();
+  TensorView* c = outputs[0]->as<TensorView>();
```

Comment:
This PR has some inconsistency here, some part of this PR tries to allow arbitrary order of fusion inputs and fusion outputs, but in `scheduleMatmul`, we are still hard coding the order. Should we have a unified scheme on how to get ABC? For example, we can have

```C++
auto mma = findMMAOp(fusion);
auto a = getAFromMMA(mma);
auto b = getBFromMMA(mma);
auto c = getCFromMMA(mma);
```

Reply:
Also, I think we need to add a test to check if different order are actually supported

Reply:
In this PR I focused on adding matmul scheduler and plugging in `schedulMatmul` to the system / segmenter. All the checks in segmenter are made so it should be reasonable to add new rules for checks when new functionality is added. This is the reason why latest revision supports an arbitrary order of inputs/outputs in the fusion - this is preparation for the support for more complex fusions.

On the other hand, I tried to avoid doing any changes in `scheduleMatmul` as the scope of this PR is already big (plugging in scheduler, compile time checks and heuristics). 

---

As for your suggestions about helpers / tools that for given fusion, will return requested input of mma from fusion. I have in my mind the following idea:
- compile time checks performs heavy analysis of dependencies in fusion:
  - if there is enough TV inputs in fusion to have at least the basic mma (2 TVs), 
  - if no more than single fusion TV input is a producer for MmaOp input
  - if there are scalar inputs (alpha/beta)
  - if there are additional TVs (out-of-place C for beta scaling)
- the product of compile-time checks are the following pieces information, that could be stored in data cache (instance of `HeuristicSummary`):
  - mapping fusion inputs / inputs' consumers to their roles in expressions in fusion, i.e. for `(alpha * A) x B` we will mark consumer of `alpha * A` as the first input for MmaOp.
  - store mma layout
- instance of heuristics params passed to `scheduleMarmul` will contain this data so we can trigger process piece by piece there. 

The major issues is that instance of `HeuristicSummary` can be initialized starting from runtime check. The stopgap solution for now would be to redo heavy analysis from compile-time check in runtime check, before any major changes are done to projects' architecture.

---

> Also, I think we need to add a test to check if different order are actually supported

I prepared a simple test like this, but without access to data from compile-time about mapping fusion inputs to mma inputs it failed in `scheduleMatmul`, on checks you have pointed out. Do you mind if implementation and test is added in the follow up PR?